wc command

shuheiokuda0704 opened the pull request.

shuheiokuda0704 committed:
Init required files include test

shuheiokuda0704 committed:
Add test (RED for now)

shuheiokuda0704 committed:
Implement wc command (not support pipe yet)

shuheiokuda0704 committed:
Add RED test for pipe

shuheiokuda0704 committed:
Support pipe

shuheiokuda0704 committed:
Fix miscount of word_num

shuheiokuda0704 committed:
Delete unnecessary code

shuheiokuda0704 committed:
Add a test for wildcard

shuheiokuda0704 committed:
Avoid directory

shuheiokuda0704 committed:
Fix rubocop warnings by `rubocop -a`

shuheiokuda0704 committed:
Fix rubocop warnings

yuuu commented on 06.wc/test/wc_command_test.rb:
```diff
@@ -0,0 +1,62 @@
+  ROOT_PATH = 'test/fixtures/sample-directory'
+
+  def test_wc_command_for_a_file
+    target_path = "#{ROOT_PATH}/aaa.txt"
```
テストが書かれていることで安心してリファクタリングができますね 👍

yuuu commented on 06.wc/lib/wc_command.rb:
```diff
@@ -0,0 +1,54 @@
+def format_items(items, line_only)
+  format_items = items.map do |item|
+    if item[:dir]
+      # format("wc: %s: read: Is a directory", item[:path])   # TODO: STDERR
```
こちら、コメントアウト＆TODOのままというのはよろしくないので何かしら対処をお願いします🙏
「ディレクトリを指定されたときにエラーを出す」という要件はないので、最悪何もしない、でもOKですので

shuheiokuda0704 replied:
標準エラー出力にメッセージを出そうとして忘れていました。対処しました

yuuu commented on 06.wc/lib/wc_command.rb:
```diff
@@ -0,0 +1,54 @@
+  paths.map do |path|
+    next { line_num: 0, word_num: 0, char_num: 0, path: path, dir: true } if File.directory?(path)
+
+    file = File.pipe?(path) ? path : File.open(path, 'r')
```
`File.open` した場合のクローズ処理が漏れているので対処をお願いします🙏

shuheiokuda0704 replied:
クローズするようにしました。

yuuu commented on 06.wc/bin/wc.rb:
```diff
@@ -0,0 +1,17 @@
+opt.on('-l') { |v| params[:line_only] = v }
+opt.parse!(ARGV)
+
+paths = ARGV.size.zero? && File.pipe?($stdin) ? [$stdin] : ARGV
```
`File.pipe?` を2箇所で判定してしまっているのですが1箇所にまとめる方法はないでしょうか？
ここで `ARGV` をそれぞれ `File.open` した結果を配列にして渡すようにすれば、`collect_items` での判定は省略できそうな気がしています。

shuheiokuda0704 replied:
試してみましたが、
collect_items 内の `{ line_num: line_num, word_num: word_num, char_num: char_num, path: File.pipe?(path) ? '' : path, dir: false }` の  `File.pipe?` をなくすうまい方法が思いついていません。
`wc.rb` 側から `run_wc` メソッドを呼び出す際にファイル名も引数で渡すのも冗長な気がしますし、もう少し考えてみます。

shuheiokuda0704 replied:
考えた結果、wc.rb の `File.pipe?` の判定は不要と判断して消し、`collect_items` メソッドの分岐をシンプルな `if ~ elsif ~` の形にしました。重複がゼロにはなってないので気になりますが、ベターにはなったかなと。。

shuheiokuda0704 committed:
Put error to STDERR if target is directory

shuheiokuda0704 committed:
Delete unnecessary `File.pipe?`

shuheiokuda0704 committed:
Refactor collect_items method

shuheiokuda0704 committed:
Fix error when no args

yuuu commented on 06.wc/lib/wc_command.rb:
```diff
@@ -8,13 +8,19 @@ def run_wc(paths:, line_only: false)
-
-    { line_num: line_num, word_num: word_num, char_num: char_num, path: File.pipe?(path) ? '' : path, dir: false }
-  end
+    if File.directory?(path)
```
修正ありがとうございます 👍 

自分としては `paths` に特性の違うもの( `IO` もしくは `文字列の配列` )が入りうる状態となっているのが複雑化の原因なのかなと思っています。

さらなる修正案としては2案、考えられます。

### 修正案1

pathsには `ARGV` をそのまま入れるようにして、 `ARGV.size.zero?` をこのメソッドで最初に判定して、 `line_num, word_num, char_num = file_stat($stdin)` とすれば全体としての複雑度やコード量は削減できそうです。

デメリットとしては `collect_items` が `$stdin` に依存するところですかね。今も `File.pipe?` を呼び出しているので実質依存している状態ではありますが。

### 修正案2

pathsには `[IO]` のみを渡すようにする、つまり、`[$stdin]` または `[File.open()]` のどちらかを渡す仕様にします。IOしか入っていないのでディレクトリでないことが確認できればそのまま `line_num, word_num, char_num = file_stat(file)` ができます。

pathを返すときに `file.tty? ? '' : file.path` のようにする必要があり、stdinのチェックが2箇所になることは防げないところですね。

shuheiokuda0704 replied:
@yuuu 
コメントありがとうございます！
修正案1 / 2 いずれもご指摘のデメリットの部分が気になって回避しました。
が、「今も `File.pipe?` を呼び出しているので実質依存している状態」なのはそのとおりなので、案1に寄せてみます。

shuheiokuda0704 committed:
Refactor
- Pass ARGV to run_wc method anyway
- Check paths' size and $stdin in collect_items method
